# Random 패키지
## 1. random 패키지의 개요
많은 프로그램과 시뮬레이션에서 무작위 번호, 즉 의사-난수(pseudo-random numbers)를 구현하는 것은 필수적\
의사 난수는 나열된 숫자들 사이에서 무작위로 추출한 것임\
난수는 여러 숫자 분포(distributions)에서 추출할 수 있음\
이 분포는 난수가 어느 범위 안에 들어와야 하는지를 결정하며, 어느 분포에 빈번하게 나타나야 하는지를 결정함\
예를 들어 random.randint 함수는 지정된 범위 내에 각 정수가 선택될 확률이 동일한 정수 값을 생성함\
공정한 6면 주사위의 규칙을 시뮬레이션할 수 있고, 각 숫자의 출현 빈도수가 약 6분의 1이 될 것으로 예상할 수 있음\
이 패키지를 사용하려면 소스 파일 시작 부분에 패키지를 탑재해야 함
```python
import random
```

## 2. random 함수 살펴보기
random 패키지는 여러 함수로 구성되었으며, 각 함수는 서로 다른 무작위 분포를 제공함
|함수 문법|설명|
|:--|:--|
|normalvariate(mean, dev)|종 곡선(bell curve)으로 알려진 고전적인 정규 분포를 생성함 <br> 높이와 너비는 다양함 <br> 인수 mean은 값의 평균값이며, 인수 dev는 표준 편차 <br> 대략적으로 값의 3분의 2 정도가 하나의 표준 편차에 포함되는 경향이 있음|
|randint(a, b)|각 정수가 선택될 확률이 동일한 a~b 범위에서 무작위 정수를 생성하며 균일한 분포임 <br> 예를 들어 randint(1, 6)은 완벽하게 공정한 육면체 주사위 결과를 시뮬레이션함|
|random()|최대 수를 제외한 0~1 범위의 무작위 부동소수점 숫자를 생성함 <br> 범위는 연속적이지만 분포가 균일하므로 N 하위 범위로 나누면 값이 각각 대략 $\frac{1}{N}$ 확률로 떨어져야 함|
|sample(population, k)|표본 모집단에서 임의로 k 원소를 생성함 <br> 모집단은 리스트, 튜플, 세트 또는 호환 가능한 컬렉션 클래스 <br> 딕셔너리에서 사용하려면 먼저 리스트로 변환해야 함|
|shuffle(list)|무작위로 리스트를 섞음 <br> 패키지의 모든 기능 중 가장 유용한 기능 <br> 반환값은 없지만, 리스트에 저장된 항목들이 섞여서 어느 위치로든 가게 됨 <br> 가령 52 카드 덱의 0부터 51까지 번호로 표현된 카드가 있을 때, shuffle(range(52))는 덱을 무작위로 섞은 리스트를 생산함| 
|uniform(a, b)|a와 b 범위에 무작위 부동소수점 숫자를 반환하며 분포는 연속적이고 균일함|

## 3. 무작위 행동 테스트하기
난수 시퀀스는 다음과 같은 특징을 가져야 함
* 대략적으로 예상 가능 : 1에서 N까지 각 값의 생성 빈도수 테스트를 여러 번 실행하면 각 값이 대략 $\frac{1}{N}$ 만큼 나올 것으로 예상해야 함

* 변동 : 값이 변할 수도 있다는 것을 예상해야 함\
10개의 균일한 값으로 100개의 실험 값을 구하는 경우 각 값이 정확히 $\frac{1}{10}$ 로 추출되기를 기대해서는 안됨\
그렇게 되면 패턴이 너무 규칙적이고 의심스러울 정도로 무작위적이지 않게 됨

* 큰 N으로 변동 감소 : 실험 값이 많아질수록 실제로 추출되는 난수의 기댓값이 실제 값에 점점 가까워진다고 기대해야 함(대수의 법칙)

이 항목들은 정량적으로 테스트하기 쉬운 것들임\
시행 횟수가 다른 테스트를 여러 번 실행하면 실제 추출 값이 예상 값에 점점 가까워지는 것을 알 수 있음\
다음 예시에서 이를 테스트하기 위해 만든 함수를 살펴보자

In [1]:
import random
def do_trials(n):
    hits = [0] * 10
    for i in range(n):
        a = random.randint(0, 9)
        hits[a] +=1
    
    for i in range(10):
        fss = '{}: {}\t {:.3}'
        print(fss.format(i, hits[i], hits[i]/(n/10)))

이 함수는 10개의 항목을 생성하면서 시작하며 각 항목은 난수의 생성 횟수를 담음\
가령 hits[0]은 0의 생성 횟수를 지정하고, hits[1]은 1의 생성 횟수를 저장하며, hits[2]는 2의 생성 횟수를 저장하는 식

첫 번째 루프는 n개의 0부터 9사이의 무작위 정수를 생성함\
그런 다음 각 숫자가 생성된 횟수를 리스트의 해당 값 항목에 누적함\
random.randint를 사용하여 균일한 확률로 0부터 9까지 범위에서 정수를 생성함

두 번째 루프는 0부터 9까지의 각 숫자가 얼마나 많이 생성되었고 각 숫자가 $\frac{1}{10}$ 씩 생성될 것이라고 예측했던 것에 비해 얼마나 생성되었는지를 보여주는 요약 결과를 출력함

In [2]:
do_trials(100)

0: 5	 0.5
1: 10	 1.0
2: 13	 1.3
3: 10	 1.0
4: 4	 0.4
5: 12	 1.2
6: 14	 1.4
7: 19	 1.9
8: 6	 0.6
9: 7	 0.7


이 100개의 실험 결과로는 난수의 생성 결과가 균일하다는 것을 설득하기에는 충분하지 않다는 것을 보여줌\
예상 생성 횟수 대비 실제 생성 횟수도 최저 0.4에서 최고 1.9로 올라감\
그러나 실험을 1000번 실행하면 조금 더 고른 결과가 나옴

In [3]:
do_trials(1000)

0: 113	 1.13
1: 92	 0.92
2: 112	 1.12
3: 92	 0.92
4: 109	 1.09
5: 93	 0.93
6: 89	 0.89
7: 103	 1.03
8: 87	 0.87
9: 110	 1.1


실행 결과를 보면 각 숫자의 실제 생성 횟수가 기대했던 $\frac{n}{10}$에 더욱 가깝다는 것을 알 수 있음\
실험 횟수를 77,000으로 증가시키면 확실히 더 가까운 결과를 얻을 수 있음

In [4]:
do_trials(77000)

0: 7646	 0.993
1: 7804	 1.01
2: 7676	 0.997
3: 7756	 1.01
4: 7471	 0.97
5: 7717	 1.0
6: 7691	 0.999
7: 7663	 0.995
8: 7751	 1.01
9: 7825	 1.02


실제 생성 횟수(7471~7825) 대비 기대 생성 횟수(전체 실험 횟수의 $\frac{1}{10}$) 비율이 3번째 열에 있음\
우리가 기대한 세 가지 난수 생성 방식을 확인하기는 충분함\
10개의 난수(0~9)는 대략 10분의 1 정도 생성되었고, 기대했던 수치와는 괴리가 있지만, 실험 횟수가 증가할수록 이 괴리가 줄어듬

## 4. 무작위-정수 게임

In [5]:
import random
n = random.randint(1, 50)
while True:
    guess = int(input('Enter guess : '))
    if guess == n:
        print('Success! You win.')
        break
    elif guess < n :
        print('Too low.', end = '')
    else:
        print('Too high.', end = '')

Too low.Too high.Too high.Too high.Too high.Too high.Success! You win.


우리는 이 게임을 몇 가지 방법으로 개선할 수 있음\
우선 게임이 끝날 때마다 사용자에게 다시 게임을 시작할 것인지 물을 수 있음\
다음으로 게임이 지루하다고 판단되면 언제든지 일찍 종료할 수 있게 할 수 있음

In [7]:
import random
def play_the_game():
    n = random.randint(1, 50)
    while True:
        guess = int(input('Enter guess (0 to exit): '))
        if guess == 0:
            print('Exiting game.')
            break
        elif guess == n:
            print('You win.')
            break
        elif guess < n:
            print('Too low.', end='')
        else:
            print('Too high.', end='')
    
while True:
    play_the_game()
    ans = input('Want to play again? (Y or N): ')
    if not ans or ans[0] in 'Nn':
        break

Too high.You win.


## 5. 카드 덱 객체 만들기
shuffle 함수는 random 패키지에서 가장 유용한 함수\
shuffle 동작 방식은 리스트 순서를 재정렬하여 항목들이 어느 위치에서나 나타날 수 있게 해줌\
항목 개수가 바뀌거나 항목이 중복되지는 않음\
가령 다음 리스트를 섞고 싶다고 가정해 보자

In [8]:
kings_list = ['John', 'James', 'Henry', 'Henry', 'George']

이제 random.shuffle로 순서를 무작위로 조정해 보자

In [10]:
random.shuffle(kings_list)
kings_list

['George', 'James', 'Henry', 'Henry', 'John']

shuffle 알고리즘은 다음 예시처럼 꽤 보편적임
```
For I in range(0, N-2),
    J = randint(I, N-1)
    Swap list[I] with list[J]
```

random.shuffle 함수의 동작 방식은 기존 항목의 값 모두를 메모리상에서 변경하지만 메모리상의 리스트를 이동시키지는 않음\
(쉽게 말하면 리스트 객체 자체는 그대로, 안에 있는 값의 순서만 바뀌는 것이며, 그래서 shuffle은 반환값이 없고 원본 리스트를 수정함)\
카드 덱의 함수를 캡슐화하는 최고의 방법 중 하나는 덱 클래스를 만들어서 덱 객체를 초기화하는데 사용하는 것임

* 0부터 N까지의 숫자를 담은 리스트를 준비하며, 52-카드 덱 안에는 각각의 유일한 숫자와 타입을 가진 카드가 포함됨

* 초기화 시 덱 객체는 스스로 숫자를 섞음(shuffle)

* 사용자는 덱 객체의 '맨 위'에서 0에서 51사이의 숫자 카드를 한 번에 하나씩 꺼냄

* 모든 카드를 꺼냈다면 덱 객체는 스스로 카드를 다시 섞음

이 프로그램이 완성되면 객체 지향 프로그램의 좋은 예시가 될 것임\
덱 객체 인스턴스는 내부 상태에서 관리될 것이며, 다음 코드는 원하는 크기로 자동으로 다시 섞이는 덱을 만들게 됨

In [15]:
import random
class Deck():
    def __init__(self, size):
        self.card_list = [i for i in range(size)]
        random.shuffle(self.card_list)
        self.current_card = 0
        self.size = size
    
    def deal(self):
        if self.size - self.current_card < 1:
            random.shuffle(self.card_list)
            self.current_card = 0
            print('Reshuffling...')
        self.current_card +=1
        return self.card_list[self.current_card - 1]

In [16]:
ranks = ['2','3','4','5','6','7','8','9','10','J','Q','K','A']
suits = ['clubs', 'diamonds', 'hearts', 'spades']
my_deck = Deck(52)

In [17]:
for i in range(12):
    for i in range(5):
        d = my_deck.deal()
        r = d % 13
        s = d // 13
        print(ranks[r], 'of', suits[s])
        print()

4 of spades

4 of diamonds

3 of diamonds

7 of hearts

J of clubs

6 of clubs

8 of diamonds

A of spades

4 of clubs

2 of hearts

K of hearts

J of hearts

Q of clubs

J of spades

K of spades

10 of clubs

7 of spades

5 of spades

J of diamonds

9 of spades

7 of clubs

K of diamonds

5 of diamonds

9 of diamonds

2 of clubs

A of hearts

A of diamonds

6 of spades

Q of diamonds

3 of spades

10 of hearts

2 of spades

10 of spades

6 of hearts

10 of diamonds

Q of spades

9 of hearts

5 of hearts

3 of hearts

9 of clubs

2 of diamonds

3 of clubs

6 of diamonds

A of clubs

8 of clubs

Q of hearts

8 of hearts

5 of clubs

K of clubs

4 of hearts

8 of spades

7 of diamonds

Reshuffling...
3 of hearts

6 of spades

K of clubs

7 of clubs

6 of diamonds

10 of hearts

4 of spades

A of spades



덱 클래스는 제약이 존재하는데, 덱이 다시 섞이는 시점에 테이블 위에 남겨진 카드가 있을 것임\
이 카드들은 섞이지 않으며 덱은 폐기된 카드만 가지고 섞어야 함

테이블 위에 남겨진 카드는 카드가 뽑힌 시점에 결정됨\
카드는 신규 차례(new hand)인 경우에만 폐기 리스트로 이동함\
다음과 같이 덱 객체를 고쳐보자

In [18]:
import random

class Deck():
    def __init__(self, size):
        self.card_list = [ i for i in range(size)]
        self.cards_in_play_list = []
        self.discards_list = []
        random.shuffle(self.card_list)
    
    def deal(self):
        if len(self.card_list) < 1:
            random.shuffle(self.discards_list)
            self.card_list = self.discards_list
            self.discards_list = []
            print('Reshuffling...')
        new_card = self.card_list.pop()
        self.cards_in_play_list.append(new_card)
        return new_card
    
    def new_hand(self):
        self.discards_list += self.cards_in_play_list
        self.cards_in_play_list.clear()

이 클래스 정의문은 1개의 신규 메서드 new_hand를 정의하고 있음\
이 신규 메서드는 현재 게임중인 카드를 모두 폐기함\
덱은 ㅇiscards_list에서 카드를 추가해야 하며, cards_in_play_list의 모든 항목은 제거(clear)됨

deal 메서드의 변경 사항을 보면, 이제는 덱 안의 모든 카드가 들어 있는 card_list를 그냥 섞는 대신 폐기된 더미만 섞음\
그 다음 섞인 카드들은 card_list로 옮겨지며 새로운 덱이 되고, 그러면 discards_list의 모든 카드는 삭제됨

테이블 위에 카드가 있을 때 카드를 섞게 되면 테이블 위의 카드는 섞이지 않기 때문에 섞은 카드의 숫자가 덱의 숫자보다 작을 것임\
그렇다면 테이블 위에 있던 카드는 현재 판이 끝나면 테이블 위에 있던 카드를 폐기하며 다시 섞여서 덱으로 돌아가게 됨